### PR TITLE
SearchKit - Support conditional links

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Advanced feature to conditionally show/hide links in a links/buttons/menu column, based on user permissions or row values.

Before
----------------------------------------
Links would display unconditionally.

After
----------------------------------------
Links can be hidden based on user permissions (e.g. "access CiviCase") or according to a value (e.g. `is_active = TRUE`).

### CODE REVIEW

File: ext/search_kit/Civi/Api4/Action/SearchDisplay/AbstractRunAction.php [hunk 1]
```diff
@@ -390,6 +393,25 @@ private function formatLinksColumn($column, $data): array {
     return $out;
   }
 
+  /**
```
Comment (eileenmcnaughton) [inline ext/search_kit/Civi/Api4/Action/SearchDisplay/AbstractRunAction.php:396]: To meet the full standard we should have a one-liner description for the function - in this case of this function it probably would make it clearer too

Non blocking

File: ext/search_kit/Civi/Api4/Action/SearchDisplay/AbstractRunAction.php [hunk 2]
```diff
@@ -390,6 +393,25 @@ private function formatLinksColumn($column, $data): array {
...
+   * @param array $item
+   * @param array $data
+   * @return bool
+   */
+  private function checkLinkCondition($item, $data): bool {
```
Comment (eileenmcnaughton) [inline ext/search_kit/Civi/Api4/Action/SearchDisplay/AbstractRunAction.php:401]: ideally type hint these params

non-blocking

**Comment by eileenmcnaughton** [comment]
@colemanw in trying to test this PR I hit a possible recent regression - going to this page

http://core-22557-5o32p.test-3.civicrm.org:8001/civicrm/acl/entityrole?action=update&id=2

And attempting to save I get this

image


My understanding is that the `CRM_Core_BAO_Managed::on_hook_civicrm_post`  function should only be called on the Managed entity - but here it is responding to the entity role

image

**Comment by colemanw** [comment]
@eileenmcnaughton it always responds to the hook (because that's the only way to implement a hook) but then decides to act based on the entity name. Why on earth is `entity = null` in that event? That's the problem.

**Comment by eileenmcnaughton** [comment]
@colemanw yeah that is fixed now - so I can try again to replicate this

File: ext/search_kit/Civi/Api4/Action/SearchDisplay/AbstractRunAction.php (hunk 1 above)
Comment (colemanw) [reply ext/search_kit/Civi/Api4/Action/SearchDisplay/AbstractRunAction.php:396]: Updated

File: ext/search_kit/Civi/Api4/Action/SearchDisplay/AbstractRunAction.php (hunk 2 above)
Comment (colemanw) [reply ext/search_kit/Civi/Api4/Action/SearchDisplay/AbstractRunAction.php]: Updated

**Comment by eileenmcnaughton** [comment]
@colemanw did you make more changes? I just tested this - it took me a bit to figure out what it was doing but once it did I added some docs

https://lab.civicrm.org/documentation/docs/user-en/-/merge_requests/548

The only 'issue' (more like  a feature gap) I noticed is that for tokens I can use fields not in the display but for the conditional links I can't. What I tried to do was add a link on the email field to the url

civicrm/activity/email/add?action=add&reset=1&selectedChild=activity&atype=3&email_id=[Contact_Email_contact_id_01.id]

but only if the on_hold field - which is not in the display = 0

Also side-issue - that url if kind of a hidden url (since normally cid rather than email id is used) but I think it would be good to define as a link in the email entity - we do define non-crud links?

**Comment by colemanw** [comment]
Thanks for the review @eileenmcnaughton. Yes I kept the code simpler and avoided making myself crazy by only supporting fields in the SELECT clause in these conditions.

**Pull request merged**